app/questionnaire/navigation.py

```diff
@@ -0,0 +1,151 @@
+import copy
+import logging
+
+from collections import defaultdict
+
+from app.data_model.answer_store import AnswerStore
+
+from app.helpers.schema_helper import SchemaHelper
+from app.questionnaire.location import Location
+from app.questionnaire.rules import evaluate_repeat
+
+logger = logging.getLogger(__name__)
+
+
+class Navigation(object):
+    """
+    Reads navigation config from the schema and returns collections of dicts that describe
+    completion status of each group
+    """
+    def __init__(self, survey_json, metadata=None, answer_store=None, completed_blocks=None):
+        self.survey_json = survey_json
+        self.metadata = metadata or {}
+        self.answer_store = answer_store or AnswerStore()
+        self.completed_blocks = completed_blocks or []
+
+    def build_navigation(self, current_group_id, current_group_instance):
+        """
+        Build navigation based on the current group/instance selected
+
+        :param current_group_id:
+        :param current_group_instance:
+        :return:
+        """
+        if 'navigation' not in self.survey_json:
+            return None
+
+        navigation = []
+
+        for group in filter(lambda x: 'hide_in_navigation' not in x, self.survey_json['groups']):
+            first_location = Location(group['id'], 0, group['blocks'][0]['id'])
+            is_current_group = group['id'] == current_group_id
+
+            default_highlight = self._should_highlight(group, current_group_id)
+
+            logger.debug("Building frontend navigation for group %s", group)
+
+            repeating_rule = SchemaHelper.get_repeat_rule(group)
+            completed_id = group['completed_id'] if 'completed_id' in group else group['blocks'][-1]['id']
+
+            if repeating_rule:
```

Comment:
This method would be more readable if we broke it out into a private method _append_repeating_navigation or navigation += _build_repeating_navigation